In [4]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Use the kagglehub client library to attach Kaggle resources like competitions, datasets, and models to your session
# Learn more about kagglehub: https://github.com/Kaggle/kagglehub/blob/main/README.md

import kagglehub
# kagglehub.dataset_download('<owner>/<dataset-slug>')

ModuleNotFoundError: No module named 'kagglehub'

In [ ]:
import os
print(os.listdir("/kaggle/input/competitions/"))

['enveda-CASMI26-molecule-id-mass-spectra']


In [8]:
import pandas as pd

train_path = "/kaggle/input/competitions/enveda-CASMI26-molecule-id-mass-spectra/train.parquet"
test_path = "/kaggle/input/competitions/enveda-CASMI26-molecule-id-mass-spectra/test.parquet"

train = pd.read_parquet(train_path)
test = pd.read_parquet(test_path)

print("Train shape:", train.shape)
print("Test shape:", test.shape)

FileNotFoundError: [Errno 2] No such file or directory: '/kaggle/input/competitions/enveda-CASMI26-molecule-id-mass-spectra/train.parquet'

In [ ]:
train.head(3)

,ingest_lib,normalized_smiles,inchikey,inchikey14,molecular_formula,ionization_mode,instrument_type,adduct,adduct_orig,precursor_mz,precursor_error_ppm,ms2_mzs,ms2_normalized_intensities,num_peaks,base_peak_intensity,collision_energy_ev,collision_energy_orig,collision_energy_orig_units
0,drug_plus,O=C1NC(=O)c2cc(Nc3ccccc3)c(Nc3ccccc3)cc21,AAALVYBICLMAMA-UHFFFAOYSA-N,AAALVYBICLMAMA,C20H15N3O2,positive,None,[M+H]+,[M+H]+,330.1237,1.671398,"[92.0495, 93.0573, 94.0607, 116.107, 123.1168,...","[0.0171037726229424, 0.328091214993177, 0.0107...",52,NaN,None,None,unknown
1,drug_plus,C#Cc1cccc(Nc2ncnc3cc(OCCOC)c(OCCOC)cc23)c1,AAKJLRGGTJKAMG-UHFFFAOYSA-N,AAKJLRGGTJKAMG,C22H23N3O4,positive,None,[M+K]+,[M+K]+,432.1320,1.302193,"[82.1451, 89.0599, 158.9638, 248.0836, 250.098...","[0.00144055962319783, 0.959302375462881, 0.5, ...",28,NaN,None,None,unknown
2,drug_plus,CC1(C)CCC(C)(C)c2cc(C(O)C(O)=Nc3ccc(C(=O)O)cc3...,AANFHDFOMFRLLR-UHFFFAOYSA-N,AANFHDFOMFRLLR,C23H26FNO4,positive,None,[M+Na]+,[M+Na]+,422.1738,1.316437,"[91.0539, 111.117, 131.0856, 150.035, 171.0803...","[0.00262159533923045, 0.0184752871374318, 0.00...",41,NaN,None,None,unknown


In [ ]:
print(train.info(show_counts=True))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2539608 entries, 0 to 2539607
Data columns (total 18 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   ingest_lib                   2539608 non-null  object 
 1   normalized_smiles            2539608 non-null  object 
 2   inchikey                     2539608 non-null  object 
 3   inchikey14                   2539608 non-null  object 
 4   molecular_formula            2539608 non-null  object 
 5   ionization_mode              2539608 non-null  object 
 6   instrument_type              2506555 non-null  object 
 7   adduct                       2539608 non-null  object 
 8   adduct_orig                  2539608 non-null  object 
 9   precursor_mz                 2539608 non-null  float64
 10  precursor_error_ppm          2535322 non-null  float64
 11  ms2_mzs                      2539608 non-null  object 
 12  ms2_normalized_intensities   2539608 non-n

In [ ]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1213 entries, 0 to 1212
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   molecule_id                  1213 non-null   object 
 1   spectrum_id                  1213 non-null   object 
 2   ms2_mzs                      1213 non-null   object 
 3   ms2_normalized_intensities   1213 non-null   object 
 4   base_peak_intensity          1213 non-null   float64
 5   adduct                       1213 non-null   object 
 6   ionization_mode              1213 non-null   object 
 7   instrument_type              1213 non-null   object 
 8   precursor_mz                 1213 non-null   float64
 9   collision_energy_orig        1213 non-null   object 
 10  collision_energy_ev          1213 non-null   object 
 11  collision_energy_orig_units  1213 non-null   object 
dtypes: float64(2), object(10)
memory usage: 113.8+ KB


In [ ]:
print("m/z:", train["ms2_mzs"].iloc[0])
print("Intensity:", train["ms2_normalized_intensities"].iloc[0])

m/z: [ 92.0495  93.0573  94.0607 116.107  123.1168 124.1203 133.0761 142.0777
 142.1226 166.0655 167.0734 168.0808 169.0838 169.0886 170.092  176.0457
 181.076  184.0758 185.0709 188.0452 192.0683 193.076  194.06   201.0533
 209.071  212.0698 219.0554 221.072  225.0532 225.0658 234.0664 234.0664
 237.066  238.0736 240.0799 250.0615 251.0687 252.0765 254.0839 257.108
 259.1231 278.0925 284.1179 285.1262 286.1341 287.1178 290.1275 301.085
 302.1284 312.1133 314.0926 328.1069]
Intensity: [1.71037726e-02 3.28091215e-01 1.07218612e-02 2.60808387e-02
 6.89810237e-03 7.31897657e-04 2.96068467e-03 1.98119487e-03
 7.14763564e-04 1.93528596e-03 2.67409777e-03 4.00310632e-02
 1.53947495e-03 4.52608905e-02 4.92054957e-03 7.90908500e-04
 3.35415059e-03 1.11231631e-02 7.63319334e-04 8.34675223e-04
 1.38966398e-03 7.90655389e-03 5.82703635e-01 3.81112746e-03
 2.94816683e-02 8.85462555e-04 1.48710610e-01 9.45472234e-04
 2.72909823e-03 1.59775040e-03 9.60284483e-03 9.60284483e-03
 1.51136640e-01 2.1635

In [ ]:
mz = train["ms2_mzs"].iloc[0]
intensity = train["ms2_normalized_intensities"].iloc[0]

print("m/z length:", len(mz))
print("Intensity length:", len(intensity))
print("Same length:", len(mz) == len(intensity))

m/z length: 52
Intensity length: 52
Same length: True


In [ ]:
print("Maximum intensity:", intensity.max())
print("Minimum intensity:", intensity.min())

Maximum intensity: 1.0
Minimum intensity: 0.00059942967118094


In [ ]:
import time
start_time = time.time()
max_intensities = train["ms2_normalized_intensities"].apply(max)

print("time taken ", time.time() - start_time)
print("Maximum of maximum intensities:", max_intensities.max())
print("Minimum of maximum intensities:", max_intensities.min())

time taken  21.100196838378906
Maximum of maximum intensities: 1.0
Minimum of maximum intensities: 1.0


In [ ]:
mz_lengths = train["ms2_mzs"].apply(len)
intensity_lengths = train["ms2_normalized_intensities"].apply(len)

print("All lengths match:", (mz_lengths == intensity_lengths).all())

All lengths match: True


In [ ]:
print(
    "All num_peaks match m/z length:",
    (train["num_peaks"] == train["ms2_mzs"].apply(len)).all()
)

All num_peaks match m/z length: True


In [ ]:
pd.set_option("display.float_format", "{:,.2f}".format)
print(train["num_peaks"].describe())

count   2,539,608.00
mean          158.22
std           516.88
min             1.00
25%            14.00
50%            42.00
75%           164.00
max        73,318.00
Name: num_peaks, dtype: float64


In [ ]:
train.columns

Index(['ingest_lib', 'normalized_smiles', 'inchikey', 'inchikey14',
       'molecular_formula', 'ionization_mode', 'instrument_type', 'adduct',
       'adduct_orig', 'precursor_mz', 'precursor_error_ppm', 'ms2_mzs',
       'ms2_normalized_intensities', 'num_peaks', 'base_peak_intensity',
       'collision_energy_ev', 'collision_energy_orig',
       'collision_energy_orig_units'],
      dtype='object')

In [ ]:
mz = train["ms2_mzs"].iloc[0]

print("Minimum m/z:", mz.min())
print("Maximum m/z:", mz.max())

Minimum m/z: 92.0495
Maximum m/z: 328.1069


In [ ]:
print(test[["molecule_id", "spectrum_id"]].head(10))

  molecule_id spectrum_id
0    m_005e53  s_4d19d522
1    m_005e53  s_5b6e28ff
2    m_005e53  s_6e514292
3    m_005e53  s_855ae19c
4    m_006153  s_3c74e9fe
5    m_006153  s_54245904
6    m_006153  s_59a80abf
7    m_006153  s_7151fa86
8    m_006153  s_86472581
9    m_006153  s_cb15f0ac


In [ ]:
test["molecule_id"]

0       m_005e53
1       m_005e53
2       m_005e53
3       m_005e53
4       m_006153
          ...   
1208    m_ff7740
1209    m_ff926d
1210    m_ff926d
1211    m_ff926d
1212    m_ff926d
Name: molecule_id, Length: 1213, dtype: object

In [ ]:
spectra_per_molecule = test.groupby("molecule_id")["spectrum_id"].count()
print(spectra_per_molecule.describe())

count   400.00
mean      3.03
std       1.38
min       1.00
25%       2.00
50%       3.00
75%       4.00
max       9.00
Name: spectrum_id, dtype: float64


In [ ]:
spectra_per_molecule_train = train.groupby("inchikey14").size()

print(spectra_per_molecule_train.describe())

count   275,810.00
mean          9.21
std          21.07
min           1.00
25%           4.00
50%           6.00
75%           9.00
max       1,468.00
dtype: float64


In [ ]:
spectra_per_molecule_train.value_counts().sort_index().head(20)

1     17033
2     16463
3     32969
4     43853
5     18741
6     19057
7     22541
8     32529
9      8803
10    10608
11     8667
12     8138
13     5313
14     3965
15     3545
16     3202
17     1722
18     1897
19     1337
20     1228
Name: count, dtype: int64

In [ ]:
train["precursor_mz"].describe()

count   2,539,608.00
mean          419.77
std         1,535.99
min             2.00
25%           308.14
50%           350.11
75%           489.21
max     2,430,333.00
Name: precursor_mz, dtype: float64

In [ ]:
train["adduct"].value_counts().head(20)

adduct
[M+H]+           1324585
[M-H]-            465688
[2M+Na]+          190033
[M+Na]+           178092
[2M+H]+            92041
[M+CH2O2-H]-       83188
[2M-H]-            56172
[M+NH4]+           53452
[M+Cl]-            27511
[M-H2O+H]+         21682
[M+K]+             18955
[M]+                8818
[M+C2H4O2-H]-       7436
[M-2H2O+H]+         3439
[M+2H]2+            1717
[2M+Na-2H]-         1431
[2M+CH2O2-H]-        665
[M-H2O]+             624
[M-CH3]-             619
[M-2H]-              454
Name: count, dtype: int64

In [ ]:
train["ionization_mode"].value_counts()

ionization_mode
positive    1899640
negative     639968
Name: count, dtype: int64

In [ ]:
train["collision_energy_ev"].head(10)

0    None
1    None
2    None
3    None
4    None
5    None
6    None
7    None
8    None
9    None
Name: collision_energy_ev, dtype: object

In [ ]:
train["collision_energy_ev"].isna().sum()

np.int64(337443)

In [ ]:
train["normalized_smiles"].head(10)

0            O=C1NC(=O)c2cc(Nc3ccccc3)c(Nc3ccccc3)cc21
1           C#Cc1cccc(Nc2ncnc3cc(OCCOC)c(OCCOC)cc23)c1
2    CC1(C)CCC(C)(C)c2cc(C(O)C(O)=Nc3ccc(C(=O)O)cc3...
3                  CN1C(=O)CN=C(c2ccccc2)c2cc(Cl)ccc21
4    NC(=O)C1=CN(C2OC(COP(=O)(O)OP(=O)(O)OCC3OC(n4c...
5                                  N=C(N)NCCN1CCCCCCC1
6    CC(C)(O)C1OC2CCC3(C)C(O)(CCC4Cc5c([nH]c6ccccc5...
7                       O=C1c2ccccc2-c2n[nH]c3cccc1c23
8    COc1cc2c(cc1OC)CC(=O)N(CCCN(C)CC1Cc3cc(OC)c(OC...
9    COC1=C(OC)C(=O)C(CC=C(C)CCC=C(C)CCC=C(C)CCC=C(...
Name: normalized_smiles, dtype: object

In [ ]:
train["inchikey14"].head(10)

0    AAALVYBICLMAMA
1    AAKJLRGGTJKAMG
2    AANFHDFOMFRLLR
3    AAOVKJBEBIDNHE
4    ACFIXJIJDZMPPO
5    ACGDKVXYNVEAGU
6    ACNHBCIZLNNLRS
7    ACPOUJIDANTYHO
8    ACRHBAYQBXXRTO
9    ACTIUHUUMQJHFO
Name: inchikey14, dtype: object

In [ ]:
train["molecular_formula"].head(10)

0       C20H15N3O2
1       C22H23N3O4
2       C23H26FNO4
3      C16H13ClN2O
4    C21H30N7O17P3
5         C10H22N4
6        C27H33NO4
7         C14H8N2O
8       C27H36N2O5
9         C59H90O4
Name: molecular_formula, dtype: object

In [ ]:
train["normalized_smiles"].duplicated().sum()

np.int64(2262042)

In [ ]:
train.groupby("normalized_smiles")["inchikey14"].nunique().max()

1

In [ ]:
train["inchikey14"].duplicated().sum()

np.int64(2263798)

In [ ]:
train["inchikey14"].nunique()

275810

In [ ]:
spectra_per_molecule_train = train.groupby("inchikey14").size()
spectra_per_molecule_train.shape

(275810,)

In [ ]:
train.groupby("inchikey14")["molecular_formula"].nunique().max()

2

In [ ]:
train.groupby("inchikey14")["molecular_formula"].nunique().value_counts().sort_index()

molecular_formula
1    275692
2       118
Name: count, dtype: int64

In [17]:
train.memory_usage(deep=True).sum() / (1024**3)

NameError: name 'train' is not defined

In [16]:
train.memory_usage(deep=True).sort_values(ascending=False)

NameError: name 'train' is not defined

In [15]:
train[["inchikey14", "ms2_mzs", "ms2_normalized_intensities"]].memory_usage(deep=True).sum() / (1024**3)

NameError: name 'train' is not defined

In [14]:
train_small = pd.read_parquet(
    train_path,
    columns=["inchikey14", "ms2_mzs", "ms2_normalized_intensities"]
)

train_small.shape

FileNotFoundError: [Errno 2] No such file or directory: '/kaggle/input/competitions/enveda-CASMI26-molecule-id-mass-spectra/train.parquet'

In [13]:
train_small.memory_usage(deep=True).sum() / (1024**3)

NameError: name 'train_small' is not defined

In [12]:
sample = pd.read_parquet(
    train_path,
    columns=["inchikey14", "ms2_mzs", "ms2_normalized_intensities"]
).head(10000)

sample.shape

FileNotFoundError: [Errno 2] No such file or directory: '/kaggle/input/competitions/enveda-CASMI26-molecule-id-mass-spectra/train.parquet'

In [ ]:
for start in range(0, 10000, 2000):
    chunk = sample.iloc[start:start+2000]
    print("Chunk:", start, "to", start + len(chunk))

Chunk: 0 to 2000
Chunk: 2000 to 4000
Chunk: 4000 to 6000
Chunk: 6000 to 8000
Chunk: 8000 to 10000


In [11]:
ranks = [1, 2, 5, 25, None]

for rank in ranks:
    if rank is None:
        score = 0
    else:
        score = 1 / rank
    print(rank, score)

1 1.0
2 0.5
5 0.2
25 0.04
None 0


In [ ]:
scores = [1.0, 0.5, 0.2, 0.04, 0]

sum(scores) / len(scores)

0.348

In [10]:
def mrr_at_25(predictions, true_molecules):
    scores = []

    for preds, true_molecule in zip(predictions, true_molecules):
        score = 0

        for rank, candidate in enumerate(preds[:25], start=1):
            if candidate == true_molecule:
                score = 1 / rank
                break

        scores.append(score)

    return sum(scores) / len(scores)

In [ ]:
predictions = [
    ["A", "B", "C", "D"],
    ["X", "Y", "Z"],
    ["P", "Q", "R"]
]

true_molecules = ["A", "Z", "M"]

mrr_at_25(predictions, true_molecules)

0.4444444444444444

In [ ]:
predictions = [
    ["A"],                    # rank 1
    ["X", "A"],               # rank 2
    ["X", "Y", "A"],          # rank 3
    ["X", "Y", "Z"],          # absent
]

true_molecules = ["A", "A", "A", "A"]

mrr_at_25(predictions, true_molecules)

0.4583333333333333

In [ ]:
preds = [f"M{i}" for i in range(1, 31)]

predictions = [preds]
true_molecules = ["M25"]

print(mrr_at_25(predictions, true_molecules))

0.04


In [ ]:
preds = [f"M{i}" for i in range(1, 31)]

predictions = [preds]
true_molecules = ["M26"]

print(mrr_at_25(predictions, true_molecules))

0.0


In [ ]:
predictions = [
    ["A", "B", "C"],
    ["X", "Y", "Z", "A"],
    ["P", "Q", "R", "S", "T"]
]

true_molecules = ["A", "A", "T"]

print(mrr_at_25(predictions, true_molecules))

0.48333333333333334


In [ ]:
test_predictions = [
    ["A", "B", "C"],
    ["X", "A", "Y"],
    ["P", "Q", "R"]
]

test_true = ["A", "A", "Z"]

test_score = mrr_at_25(test_predictions, test_true)

print("Test MRR@25:", test_score)

Test MRR@25: 0.5


# Day 9 — MRR@25 validation
#
# Tested:
# 1. Rank 1  -> 1.0
# 2. Rank 2  -> 0.5
# 3. Rank 25 -> 0.04
# 4. Rank >25 -> 0
# 5. Multiple molecules
#
# Result: scorer behaves as expected.

In [ ]:
train["inchikey14"].nunique()

275810

In [ ]:
from sklearn.model_selection import train_test_split

molecules = train["inchikey14"].unique()

train_molecules, val_molecules = train_test_split(
    molecules,
    test_size=0.20,
    random_state=42
)

len(train_molecules), len(val_molecules)

(220648, 55162)

In [ ]:
set(train_molecules).intersection(set(val_molecules))

set()

In [18]:
train_df = train[train["inchikey14"].isin(train_molecules)]
val_df = train[train["inchikey14"].isin(val_molecules)]

train_df.shape, val_df.shape

NameError: name 'train' is not defined

In [19]:
train_df["inchikey14"].nunique(), val_df["inchikey14"].nunique()

NameError: name 'train_df' is not defined

In [20]:
overlap = set(train_df["inchikey14"]).intersection(
    set(val_df["inchikey14"])
)

len(overlap)

NameError: name 'train_df' is not defined

In [21]:
train_spectra = set(
    zip(
        train_df["ms2_mzs"].astype(str),
        train_df["ms2_normalized_intensities"].astype(str)
    )
)

val_spectra = set(
    zip(
        val_df["ms2_mzs"].astype(str),
        val_df["ms2_normalized_intensities"].astype(str)
    )
)

len(train_spectra.intersection(val_spectra))

NameError: name 'train_df' is not defined

In [ ]:
train_df[["inchikey14", "ms2_mzs", "ms2_normalized_intensities"]].head()

In [9]:
train_sample = train_df.head(100000).copy()
val_sample = val_df.head(100000).copy()

train_hashes = train_sample["ms2_mzs"].astype(str).map(hash)
val_hashes = val_sample["ms2_mzs"].astype(str).map(hash)

len(set(train_hashes) & set(val_hashes))

NameError: name 'train_df' is not defined

In [ ]:
train_sample = train_df.head(100000).copy()
val_sample = val_df.head(100000).copy()

train_sample["spectrum_hash"] = (
    train_sample["ms2_mzs"].astype(str) + "|" +
    train_sample["ms2_normalized_intensities"].astype(str)
).map(hash)

val_sample["spectrum_hash"] = (
    val_sample["ms2_mzs"].astype(str) + "|" +
    val_sample["ms2_normalized_intensities"].astype(str)
).map(hash)

In [ ]:
common_hashes = set(train_sample["spectrum_hash"]) & set(val_sample["spectrum_hash"])

train_matches = train_sample[train_sample["spectrum_hash"].isin(common_hashes)]
val_matches = val_sample[val_sample["spectrum_hash"].isin(common_hashes)]

train_matches[["inchikey14", "spectrum_hash"]], val_matches[["inchikey14", "spectrum_hash"]]

(          inchikey14        spectrum_hash
 7493  ACRZDVQTYHTYRF -4165698741540319317
 7494  ACRZDVQTYHTYRF -2417563347741688487
 7495  ACRZDVQTYHTYRF  2991433399413113910,
             inchikey14        spectrum_hash
 164426  DJWGNJBHNRDPKL  2991433399413113910
 164427  DJWGNJBHNRDPKL -4165698741540319317
 164428  DJWGNJBHNRDPKL -2417563347741688487)

# Day 11 — Leakage Check Finding
#
# No molecule overlap exists between train and validation.
# However, 3 identical complete spectra were found in the 100k-row
# train/validation samples.
#
# These identical spectra were associated with different inchikey14
# values, showing that identical spectrum data can occur for different
# molecules in this dataset.
#
# This finding may be relevant later when evaluating spectral retrieval
# and candidate ranking.

In [ ]:
!pip install rdkit

ERROR: Could not find a version that satisfies the requirement rdkit (from versions: none)
ERROR: No matching distribution found for rdkit


In [ ]:
import sys
print(sys.version)

!conda list | grep -i rdkit

3.12.13 (main, Mar  4 2026, 09:23:07) [GCC 11.4.0]
/bin/bash: line 1: conda: command not found


In [ ]:
from rdkit import Chem
from rdkit.Chem import AllChem

print("RDKit working")

In [ ]:
import sys
print(sys.version)

3.12.14 | packaged by Anaconda, Inc. | (main, Aug 27 2026, 14:37:13) [MSC v.1942 64 bit (AMD64)]


In [ ]:
conda install -c conda-forge rdkit

3 channel Terms of Service accepted
Retrieving notices: done
Channels:
 - conda-forge
 - defaults
Platform: win-64
Solving environment: done

## Package Plan ##

  environment location: c:\Users\Lenovo\anaconda3\envs\ml

  added / updated specs:
    - rdkit


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    _python_abi3_support-1.0   |       hd8ed1ab_3           8 KB  conda-forge
    charset-normalizer-3.5.2   |       h308a848_0         171 KB  conda-forge
    cpython-3.12.14            |  py312hd8ed1ab_3          46 KB  conda-forge
    freetype-py-2.3.0          |     pyhd8ed1ab_0          58 KB  conda-forge
    libboost-1.86.0            |       h9dfe17d_5         2.3 MB  conda-forge
    libboost-python-1.86.0     |  py312h9b46583_5         109 KB  conda-forge
    libfreetype-2.14.1         |       h57928b3_0           8 KB  conda-forge
    libfreetype6-2.14.1        |       hdbac1cb_0 



==> WARNING: A newer version of conda exists. <==
    current version: 25.5.1
    latest version: 26.9.1

Please update conda by running

    $ conda update -n base -c defaults conda




In [ ]:
from rdkit import Chem
from rdkit.Chem import AllChem, DataStructs

print("RDKit working")

RDKit working


In [ ]:
import pandas as pd

train = pd.read_parquet(
    r"E:\project\Kaggle\data\train.parquet",
    columns=["normalized_smiles", "inchikey14"]
)

print(train.shape)

(2539608, 2)


In [ ]:
smiles = train["normalized_smiles"].iloc[0]

mol = Chem.MolFromSmiles(smiles)

print(smiles)
print(mol is not None)

O=C1NC(=O)c2cc(Nc3ccccc3)c(Nc3ccccc3)cc21
True


In [ ]:
from rdkit.Chem import rdFingerprintGenerator

generator = rdFingerprintGenerator.GetMorganGenerator(radius=2, fpSize=2048)

fp = generator.GetFingerprint(mol)

print(fp.GetNumBits())
print(fp.GetNumOnBits())

2048
25


In [ ]:
unique_molecules = train[["inchikey14", "normalized_smiles"]].drop_duplicates()

print(unique_molecules.shape)

(277566, 2)


In [ ]:
smiles_per_molecule = unique_molecules.groupby("inchikey14")["normalized_smiles"].nunique()

print(smiles_per_molecule.value_counts().sort_index())

normalized_smiles
1    274198
2      1494
3        97
4        16
5         5
Name: count, dtype: int64


In [ ]:
print(unique_molecules[unique_molecules["inchikey14"] == unique_molecules["inchikey14"].value_counts().idxmax()])

             inchikey14                                  normalized_smiles
113      BAWFJGJZGIEFAR  NC(=O)c1ccc[n+](C2OC(COP(=O)(O)OP(=O)(O)OCC3OC...
1165821  BAWFJGJZGIEFAR  NC(=O)c1ccc[n+](C2OC(COP(=O)([O-])OP(=O)(O)OCC...
1165823  BAWFJGJZGIEFAR  N=C(O)c1ccc[n+](C2OC(COP(=O)(O)OP(=O)([O-])OCC...
1165842  BAWFJGJZGIEFAR  N=C(O)c1ccc[n+](C2OC(COP(=O)(O)OP(=O)(O)OCC3OC...
1483555  BAWFJGJZGIEFAR  NC(=O)c1ccc[n+](C2OC(COP(=O)(O)OP(=O)([O-])OCC...


In [ ]:
molecules = (
    unique_molecules
    .drop_duplicates("inchikey14")
    .reset_index(drop=True)
)

print(molecules.shape)

(275810, 2)


In [ ]:
from rdkit.Chem import rdFingerprintGenerator

generator = rdFingerprintGenerator.GetMorganGenerator(
    radius=2,
    fpSize=2048
)

molecules["fingerprint"] = molecules["normalized_smiles"].apply(
    lambda x: generator.GetFingerprint(Chem.MolFromSmiles(x))
)

print(molecules.shape)

(275810, 3)


In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split

molecule_ids = np.array(molecules["inchikey14"].tolist())

train_ids, val_ids = train_test_split(
    molecule_ids,
    test_size=0.20,
    random_state=42
)

print(len(train_ids))
print(len(val_ids))

220648
55162


In [ ]:
train_molecules = molecules[molecules["inchikey14"].isin(train_ids)]
val_molecules = molecules[molecules["inchikey14"].isin(val_ids)]

print(train_molecules.shape)
print(val_molecules.shape)

(220648, 3)
(55162, 3)


In [ ]:
from rdkit import DataStructs

val_fp = val_molecules["fingerprint"].iloc[0]

similarities = DataStructs.BulkTanimotoSimilarity(
    val_fp,
    train_molecules["fingerprint"].tolist()
)

print(max(similarities))

0.88


In [ ]:
sample_val = val_molecules.head(100)

max_similarities = []

for fp in sample_val["fingerprint"]:
    similarities = DataStructs.BulkTanimotoSimilarity(
        fp,
        train_molecules["fingerprint"].tolist()
    )
    max_similarities.append(max(similarities))

print(max_similarities[:10])

[0.88, 0.6623376623376623, 0.8032786885245902, 0.7169811320754716, 0.3548387096774194, 0.7049180327868853, 0.7037037037037037, 0.9591836734693877, 0.3548387096774194, 0.8571428571428571]


In [ ]:
import numpy as np

print("Minimum:", min(max_similarities))
print("Median:", np.median(max_similarities))
print("Maximum:", max(max_similarities))

Minimum: 0.26785714285714285
Median: 0.6759259259259259
Maximum: 1.0


In [ ]:
sample_val = val_molecules.head(1000)

max_similarities = []

for fp in sample_val["fingerprint"]:
    similarities = DataStructs.BulkTanimotoSimilarity(
        fp,
        train_molecules["fingerprint"].tolist()
    )
    max_similarities.append(max(similarities))

print("Minimum:", min(max_similarities))
print("Median:", np.median(max_similarities))
print("Maximum:", max(max_similarities))

Minimum: 0.23809523809523808
Median: 0.6440677966101694
Maximum: 1.0


In [ ]:
import numpy as np

threshold_low = np.percentile(max_similarities, 20)
threshold_high = np.percentile(max_similarities, 80)

print("20th percentile:", threshold_low)
print("80th percentile:", threshold_high)

20th percentile: 0.5415300546448087
80th percentile: 0.7648897058823529


In [ ]:
# Day 12 — Known/Unseen Validation
#
# Used RDKit Morgan fingerprints and Tanimoto similarity.
# For each validation molecule, maximum similarity to the training
# molecule set represents how similar it is to the training chemistry.
#
# On a 1,000-molecule validation sample:
# Minimum similarity: 0.238
# Median similarity: 0.644
# Maximum similarity: 1.000
#
# Sample-based percentile boundaries:
# 20th percentile: 0.542
# 80th percentile: 0.765
#
# Proposed validation subsets:
# Bottom 20%  -> unseen-style
# Middle 60%  -> general
# Top 20%     -> known-style
#
# Full 55,162-molecule similarity calculation can be performed later
# if needed for final validation experiments.

In [1]:
import pyarrow.parquet as pq

parquet_path = r"E:\project\Kaggle\data\train.parquet"

pf = pq.ParquetFile(parquet_path)

print("Number of row groups:", pf.num_row_groups)
print("Rows in first row group:", pf.metadata.row_group(0).num_rows)

Number of row groups: 21
Rows in first row group: 131072


In [2]:
train_sample = pf.read_row_group(
    0,
    columns=[
        "inchikey14",
        "normalized_smiles",
        "ms2_mzs",
        "ms2_normalized_intensities"
    ]
).to_pandas()

print(train_sample.shape)

(131072, 4)


In [5]:
query = train_sample.iloc[0]

print("True molecule:", query["inchikey14"])
print("Number of peaks:", len(query["ms2_mzs"]))

NameError: name 'train_sample' is not defined

In [5]:
candidates = train_sample["inchikey14"].drop_duplicates().sample(
    n=25,
    random_state=42
).tolist()

print(len(candidates))
print(candidates)

25
['CBQRNEGLHQARTO', 'AZZNYCNGOHWCDR', 'AYUZBGWDQKFOIX', 'BPKPVJMFTTZOAA', 'BNTKUVZVWYLCOH', 'BZHLPUSMHWGMTB', 'BAQJPUAPRWRUQG', 'ACIBKVSURJRHTE', 'CQOLXTILTOHVCM', 'BJSMJGMGNZQRTO', 'BAVVXCOFWGUXKC', 'WVVSZNPYNCNODU', 'BTMOSQQEMVDEPV', 'AIJIIAWOJCILRI', 'ANBYBYGSSJXEMA', 'AFBFFLVQQXWPHS', 'BWYULXYRNMKSNP', 'BGDKAVGWHJFAGW', 'CRKPYCIHXNZABY', 'BPPOEAWIQTZLFL', 'CNJZDJVIIKDHQJ', 'ATLHDZNQHLIHLG', 'BOMYPWFCPYQYJB', 'CLTKLCFPEVPKBB', 'AGPSLSAMWLJKKI']


In [12]:
import random

candidate_scores = {
    candidate: random.random()
    for candidate in candidates
}

ranked_candidates = sorted(
    candidate_scores,
    key=candidate_scores.get,
    reverse=True
)

print(ranked_candidates)

['CBQRNEGLHQARTO', 'CLTKLCFPEVPKBB', 'BAVVXCOFWGUXKC', 'CRKPYCIHXNZABY', 'CNJZDJVIIKDHQJ', 'WVVSZNPYNCNODU', 'AZZNYCNGOHWCDR', 'BGDKAVGWHJFAGW', 'AIJIIAWOJCILRI', 'ANBYBYGSSJXEMA', 'BTMOSQQEMVDEPV', 'CQOLXTILTOHVCM', 'BNTKUVZVWYLCOH', 'BJSMJGMGNZQRTO', 'AYUZBGWDQKFOIX', 'BZHLPUSMHWGMTB', 'ATLHDZNQHLIHLG', 'AGPSLSAMWLJKKI', 'ACIBKVSURJRHTE', 'BOMYPWFCPYQYJB', 'BPKPVJMFTTZOAA', 'BWYULXYRNMKSNP', 'BAQJPUAPRWRUQG', 'AFBFFLVQQXWPHS', 'BPPOEAWIQTZLFL']


In [3]:
true_molecule = query["inchikey14"]

if true_molecule in ranked_candidates:
    rank = ranked_candidates.index(true_molecule) + 1
else:
    rank = None

print("True molecule:", true_molecule)
print("Rank:", rank)

NameError: name 'query' is not defined

In [14]:
score = mrr_at_25(
    [ranked_candidates],
    [true_molecule]
)

print("MRR@25:", score)

MRR@25: 0.0


In [ ]:
# Day 13 — Tiny End-to-End Dummy Prediction Pipeline
#
# Built a minimal pipeline:
# spectrum → candidate generation → ranking → top-25 prediction → MRR@25
#
# Candidate generation used random molecules.
# Result: true molecule was not in top-25.
# MRR@25 = 0.0
#
# Purpose: verify that the complete prediction/evaluation pipeline works.
# This is a dummy baseline, not a meaningful prediction method.

In [15]:
validation_sample = train_sample.head(100).copy()

print(validation_sample.shape)

(100, 4)


In [4]:
all_predictions = []
true_molecules = []

for _, query in validation_sample.iterrows():

    candidates = train_sample["inchikey14"].drop_duplicates().sample(
        n=25,
        random_state=42
    ).tolist()

    all_predictions.append(candidates)
    true_molecules.append(query["inchikey14"])

score = mrr_at_25(
    all_predictions,
    true_molecules
)

print("Validation MRR@25:", score)

NameError: name 'validation_sample' is not defined

In [ ]:
# Day 14 — First Validation Sanity Check
#
# Validation samples: 100
# Candidate generation: random 25 molecules
# Validation MRR@25: 0.0
#
# Conclusion:
# The end-to-end pipeline runs correctly.
# Random candidate generation provides no useful retrieval performance.
# This establishes the dummy baseline before implementing real
# spectral retrieval.

In [17]:
print(len(all_predictions))
print(len(true_molecules))
print(all(len(x) == 25 for x in all_predictions))

100
100
True


In [23]:
import pyarrow.parquet as pq

parquet_path = r"E:\project\Kaggle\data\train.parquet"

pf = pq.ParquetFile(parquet_path)

train_sample = pf.read_row_group(
    0,
    columns=[
        "inchikey14",
        "normalized_smiles",
        "ms2_mzs",
        "ms2_normalized_intensities"
    ]
).to_pandas()

print(train_sample.shape)

(131072, 4)


In [ ]:
query = train_sample.iloc[0]

mzs = query["ms2_mzs"]
intensities = query["ms2_normalized_intensities"]

print("Number of peaks:", len(mzs))
print("First 10 m/z:", mzs[:10])
print("First 10 intensities:", intensities[:10])

Number of peaks: 52
First 10 m/z: [ 92.0495  93.0573  94.0607 116.107  123.1168 124.1203 133.0761 142.0777
 142.1226 166.0655]
First 10 intensities: [0.01710377 0.32809121 0.01072186 0.02608084 0.0068981  0.0007319
 0.00296068 0.00198119 0.00071476 0.00193529]


In [25]:
print("Sorted by m/z:", all(mzs[i] <= mzs[i+1] for i in range(len(mzs)-1)))

Sorted by m/z: True


In [26]:
print("Zero intensity peaks:", sum(i == 0 for i in intensities))
print("Negative intensity peaks:", sum(i < 0 for i in intensities))

Zero intensity peaks: 0
Negative intensity peaks: 0


In [27]:
import numpy as np

print("Invalid m/z:", np.sum(~np.isfinite(mzs)))
print("Invalid intensity:", np.sum(~np.isfinite(intensities)))

Invalid m/z: 0
Invalid intensity: 0


In [28]:
import numpy as np

def clean_spectrum(mzs, intensities):
    mzs = np.asarray(mzs, dtype=float)
    intensities = np.asarray(intensities, dtype=float)

    # Keep finite and positive peaks
    mask = (
        np.isfinite(mzs) &
        np.isfinite(intensities) &
        (intensities > 0)
    )

    mzs = mzs[mask]
    intensities = intensities[mask]

    # Sort by m/z
    order = np.argsort(mzs)

    return mzs[order], intensities[order]

In [29]:
clean_mzs, clean_intensities = clean_spectrum(mzs, intensities)

print("Original peaks:", len(mzs))
print("Cleaned peaks:", len(clean_mzs))
print("First 10 cleaned m/z:", clean_mzs[:10])

Original peaks: 52
Cleaned peaks: 52
First 10 cleaned m/z: [ 92.0495  93.0573  94.0607 116.107  123.1168 124.1203 133.0761 142.0777
 142.1226 166.0655]


In [30]:
test_mzs = np.array([150, 100, 200])
test_intensities = np.array([0.5, 0, 0.8])

clean_test_mzs, clean_test_intensities = clean_spectrum(
    test_mzs, test_intensities
)

print(clean_test_mzs)
print(clean_test_intensities)

[150. 200.]
[0.5 0.8]


In [ ]:
for i in range(10):
    mzs = train_sample.iloc[i]["ms2_mzs"]
    intensities = train_sample.iloc[i]["ms2_normalized_intensities"]

    clean_mzs, clean_intensities = clean_spectrum(mzs, intensities)

    print(
        f"Spectrum {i}: "
        f"{len(mzs)} → {len(clean_mzs)} peaks"
    )

Spectrum 0: 52 → 52 peaks
Spectrum 1: 28 → 28 peaks
Spectrum 2: 41 → 41 peaks
Spectrum 3: 166 → 166 peaks
Spectrum 4: 186 → 186 peaks
Spectrum 5: 119 → 119 peaks
Spectrum 6: 200 → 200 peaks
Spectrum 7: 9 → 9 peaks
Spectrum 8: 55 → 55 peaks
Spectrum 9: 200 → 200 peaks


In [32]:
for i in range(10):
    mzs = train_sample.iloc[i]["ms2_mzs"]
    intensities = train_sample.iloc[i]["ms2_normalized_intensities"]

    print(i, len(mzs), len(intensities), len(mzs) == len(intensities))

0 52 52 True
1 28 28 True
2 41 41 True
3 166 166 True
4 186 186 True
5 119 119 True
6 200 200 True
7 9 9 True
8 55 55 True
9 200 200 True


In [33]:
changed = 0

for i in range(1000):
    mzs = train_sample.iloc[i]["ms2_mzs"]
    intensities = train_sample.iloc[i]["ms2_normalized_intensities"]

    clean_mzs, clean_intensities = clean_spectrum(mzs, intensities)

    if len(clean_mzs) != len(mzs):
        changed += 1

print("Spectra changed after cleaning:", changed)

Spectra changed after cleaning: 0


# Day 16 conclusion:
# In a sample of 1,000 spectra, basic cleaning removed 0 peaks.
# m/z values were already sorted and m/z/intensity lengths matched.
# Therefore, no aggressive cleaning or intensity threshold is applied.

In [34]:
mzs = train_sample.iloc[0]["ms2_mzs"]
intensities = np.array(
    train_sample.iloc[0]["ms2_normalized_intensities"],
    dtype=float
)

print("Min:", intensities.min())
print("Max:", intensities.max())
print("First 10:", intensities[:10])

Min: 0.00059942967118094
Max: 1.0
First 10: [0.01710377 0.32809121 0.01072186 0.02608084 0.0068981  0.0007319
 0.00296068 0.00198119 0.00071476 0.00193529]


In [35]:
sqrt_intensities = np.sqrt(intensities)

log_intensities = np.log1p(intensities)

print("Original:", intensities[:10])
print("Square root:", sqrt_intensities[:10])
print("Log:", log_intensities[:10])

Original: [0.01710377 0.32809121 0.01072186 0.02608084 0.0068981  0.0007319
 0.00296068 0.00198119 0.00071476 0.00193529]
Square root: [0.13078139 0.57279247 0.10354642 0.16149563 0.08305482 0.02705361
 0.05441217 0.04451062 0.02673506 0.04399189]
Log: [0.01695915 0.28374273 0.01066479 0.02574653 0.00687442 0.00073163
 0.00295631 0.00197923 0.00071451 0.00193342]


In [36]:
print("Original :", intensities.min(), intensities.max())
print("Sqrt     :", sqrt_intensities.min(), sqrt_intensities.max())
print("Log      :", log_intensities.min(), log_intensities.max())

Original : 0.00059942967118094 1.0
Sqrt     : 0.024483252871727236 1.0
Log      : 0.0005992500849782092 0.6931471805599453


In [37]:
print("Original order:", np.argsort(intensities)[::-1][:10])
print("Sqrt order:    ", np.argsort(sqrt_intensities)[::-1][:10])
print("Log order:     ", np.argsort(log_intensities)[::-1][:10])

Original order: [45 49 22  1 33 37 32 26 36 13]
Sqrt order:     [45 49 22  1 33 37 32 26 36 13]
Log order:      [45 49 22  1 33 37 32 26 36 13]


In [38]:
print("Original mean:", intensities.mean())
print("Sqrt mean:", sqrt_intensities.mean())
print("Log mean:", log_intensities.mean())

Original mean: 0.06950932931120447
Sqrt mean: 0.15699330361117853
Log mean: 0.056455308304938126


In [39]:
mzs_A = train_sample.iloc[0]["ms2_mzs"]
int_A = np.array(train_sample.iloc[0]["ms2_normalized_intensities"])

mzs_B = train_sample.iloc[1]["ms2_mzs"]
int_B = np.array(train_sample.iloc[1]["ms2_normalized_intensities"])

print(len(mzs_A), len(mzs_B))

52 28


In [40]:
import numpy as np

bin_width = 1.0

min_mz = min(mzs_A.min(), mzs_B.min())
max_mz = max(mzs_A.max(), mzs_B.max())

bins = np.arange(
    np.floor(min_mz),
    np.ceil(max_mz) + bin_width,
    bin_width
)

print("Number of bins:", len(bins) - 1)
print("First 10 bins:", bins[:10])

Number of bins: 303
First 10 bins: [82. 83. 84. 85. 86. 87. 88. 89. 90. 91.]


In [41]:
A_binned = np.zeros(len(bins) - 1)

bin_indices_A = np.digitize(mzs_A, bins) - 1

for idx, intensity in zip(bin_indices_A, int_A):
    A_binned[idx] += intensity

print("Non-zero bins in A:", np.count_nonzero(A_binned))

Non-zero bins in A: 48


In [42]:
B_binned = np.zeros(len(bins) - 1)

bin_indices_B = np.digitize(mzs_B, bins) - 1

for idx, intensity in zip(bin_indices_B, int_B):
    B_binned[idx] += intensity

print("Non-zero bins in B:", np.count_nonzero(B_binned))

Non-zero bins in B: 27


In [43]:
from sklearn.metrics.pairwise import cosine_similarity

original_similarity = cosine_similarity(
    A_binned.reshape(1, -1),
    B_binned.reshape(1, -1)
)[0, 0]

print("Original cosine similarity:", original_similarity)

Original cosine similarity: 0.003704431663427373


In [44]:
A_sqrt = np.sqrt(A_binned)
B_sqrt = np.sqrt(B_binned)

sqrt_similarity = cosine_similarity(
    A_sqrt.reshape(1, -1),
    B_sqrt.reshape(1, -1)
)[0, 0]

print("Sqrt cosine similarity:", sqrt_similarity)

Sqrt cosine similarity: 0.03327199953208235


In [45]:
A_log = np.log1p(A_binned)
B_log = np.log1p(B_binned)

log_similarity = cosine_similarity(
    A_log.reshape(1, -1),
    B_log.reshape(1, -1)
)[0, 0]

print("Log cosine similarity:", log_similarity)

Log cosine similarity: 0.004880271270390677


In [46]:
pairs = [(i, i + 1) for i in range(100)]

print("Number of pairs:", len(pairs))
print("First 5 pairs:", pairs[:5])

Number of pairs: 100
First 5 pairs: [(0, 1), (1, 2), (2, 3), (3, 4), (4, 5)]


In [48]:
results = []

for i, j in pairs:
    mzs_A = train_sample.iloc[i]["ms2_mzs"]
    int_A = np.array(train_sample.iloc[i]["ms2_normalized_intensities"])

    mzs_B = train_sample.iloc[j]["ms2_mzs"]
    int_B = np.array(train_sample.iloc[j]["ms2_normalized_intensities"])

    # Common bins
    min_mz = min(mzs_A.min(), mzs_B.min())
    max_mz = max(mzs_A.max(), mzs_B.max())

    bins = np.arange(
        np.floor(min_mz),
        np.ceil(max_mz) + 1,
        1.0
    )

    # Bin A
    A_binned = np.zeros(len(bins) - 1)
    idx_A = np.digitize(mzs_A, bins) - 1
    idx_A = np.clip(idx_A, 0, len(A_binned) - 1)
    for idx, intensity in zip(idx_A, int_A):
        A_binned[idx] += intensity

    # Bin B
    B_binned = np.zeros(len(bins) - 1)
    idx_B = np.digitize(mzs_B, bins) - 1
    idx_B = np.clip(idx_B, 0, len(B_binned) - 1)
    for idx, intensity in zip(idx_B, int_B):
        B_binned[idx] += intensity

    # Similarities
    original = cosine_similarity(
        A_binned.reshape(1, -1),
        B_binned.reshape(1, -1)
    )[0, 0]

    sqrt = cosine_similarity(
        np.sqrt(A_binned).reshape(1, -1),
        np.sqrt(B_binned).reshape(1, -1)
    )[0, 0]

    log = cosine_similarity(
        np.log1p(A_binned).reshape(1, -1),
        np.log1p(B_binned).reshape(1, -1)
    )[0, 0]

    results.append((original, sqrt, log))

print("Completed:", len(results))

Completed: 100


In [ ]:
results = np.array(results)

print("Original mean:", results[:, 0].mean())
print("Sqrt mean:", results[:, 1].mean())
print("Log mean:", results[:, 2].mean())

Original mean: 0.036986600648806045
Sqrt mean: 0.11286402545546473
Log mean: 0.044779851167499496


On 100 adjacent training-spectrum pairs using 1 m/z binning, √I produced the highest mean cosine similarity.”

In [50]:
rng = np.random.default_rng(42)

random_pairs = [
    tuple(rng.choice(len(train_sample), size=2, replace=False))
    for _ in range(100)
]

print("First 5 random pairs:", random_pairs[:5])

First 5 random pairs: [(np.int64(11698), np.int64(101443)), (np.int64(57524), np.int64(56756)), (np.int64(91405), np.int64(11264)), (np.int64(12343), np.int64(69006)), (np.int64(96435), np.int64(99764))]


In [51]:
results = []

for i, j in random_pairs:
    mzs_A = train_sample.iloc[i]["ms2_mzs"]
    int_A = np.array(train_sample.iloc[i]["ms2_normalized_intensities"])

    mzs_B = train_sample.iloc[j]["ms2_mzs"]
    int_B = np.array(train_sample.iloc[j]["ms2_normalized_intensities"])

    # Common bins
    min_mz = min(mzs_A.min(), mzs_B.min())
    max_mz = max(mzs_A.max(), mzs_B.max())

    bins = np.arange(
        np.floor(min_mz),
        np.ceil(max_mz) + 1,
        1.0
    )

    # Bin A
    A_binned = np.zeros(len(bins) - 1)
    idx_A = np.digitize(mzs_A, bins) - 1
    idx_A = np.clip(idx_A, 0, len(A_binned) - 1)
    for idx, intensity in zip(idx_A, int_A):
        A_binned[idx] += intensity

    # Bin B
    B_binned = np.zeros(len(bins) - 1)
    idx_B = np.digitize(mzs_B, bins) - 1
    idx_B = np.clip(idx_B, 0, len(B_binned) - 1)
    for idx, intensity in zip(idx_B, int_B):
        B_binned[idx] += intensity

    # Similarities
    original = cosine_similarity(
        A_binned.reshape(1, -1),
        B_binned.reshape(1, -1)
    )[0, 0]

    sqrt = cosine_similarity(
        np.sqrt(A_binned).reshape(1, -1),
        np.sqrt(B_binned).reshape(1, -1)
    )[0, 0]

    log = cosine_similarity(
        np.log1p(A_binned).reshape(1, -1),
        np.log1p(B_binned).reshape(1, -1)
    )[0, 0]

    results.append((original, sqrt, log))

print("Completed:", len(results))

Completed: 100


In [52]:
random_results = np.array(results)

print("Original mean:", random_results[:, 0].mean())
print("Sqrt mean:", random_results[:, 1].mean())
print("Log mean:", random_results[:, 2].mean())

Original mean: 0.01043704375315987
Sqrt mean: 0.05945041251846898
Log mean: 0.013303093722925484


√I is the most promising intensity transform among the three tested.

In [53]:
mz_tolerance = 0.01

In [54]:
def count_peak_matches(mzs_A, mzs_B, tolerance=0.01):
    matches = 0

    for mz_A in mzs_A:
        if np.any(np.abs(mzs_B - mz_A) <= tolerance):
            matches += 1

    return matches

In [55]:
matches = count_peak_matches(mzs_A, mzs_B)

print("Matching peaks:", matches)

Matching peaks: 16


In [57]:
similarity = matches / min(len(mzs_A), len(mzs_B))

print("Peak-matching similarity:", similarity)

Peak-matching similarity: 0.10526315789473684


In [58]:
print("matches:", matches)
print("len A:", len(mzs_A))
print("len B:", len(mzs_B))
print("min length:", min(len(mzs_A), len(mzs_B)))

matches: 16
len A: 152
len B: 213
min length: 152


In [59]:
intensities_A = train_sample.iloc[0]["ms2_normalized_intensities"]
intensities_B = train_sample.iloc[1]["ms2_normalized_intensities"]

matched_intensities = []

for mz_A, intensity_A in zip(mzs_A, intensities_A):
    distances = np.abs(mzs_B - mz_A)
    
    if np.min(distances) <= 0.01:
        matched_intensities.append(intensity_A)

print("Number of matched intensities:", len(matched_intensities))
print(matched_intensities[:10])

Number of matched intensities: 9
[np.float64(0.328091214993177), np.float64(0.00296068467031701), np.float64(0.00153947494653074), np.float64(0.0452608905384016), np.float64(0.00492054957457935), np.float64(0.0111231630618452), np.float64(0.00272909822539696), np.float64(0.0622143601635948), np.float64(0.00266525116771722)]


In [60]:
for mz_A, intensity_A in zip(mzs_A, intensities_A):
    distances = np.abs(mzs_B - mz_A)
    
    if np.min(distances) <= 0.01:
        print(
            "A m/z:", mz_A,
            "B nearest m/z:", mzs_B[np.argmin(distances)],
            "difference:", np.min(distances)
        )

A m/z: 66.0487 B nearest m/z: 66.0436 difference: 0.005099999999998772
A m/z: 75.0251 B nearest m/z: 75.0231 difference: 0.001999999999995339
A m/z: 82.0302 B nearest m/z: 82.03 difference: 0.00019999999999242846
A m/z: 82.0674 B nearest m/z: 82.0647 difference: 0.0027000000000043656
A m/z: 83.0507 B nearest m/z: 83.048 difference: 0.0027000000000043656
A m/z: 91.0553 B nearest m/z: 91.0532 difference: 0.0020999999999986585
A m/z: 94.0656 B nearest m/z: 94.0653 difference: 0.00030000000000995897
A m/z: 106.0661 B nearest m/z: 106.0641 difference: 0.0020000000000095497
A m/z: 110.0558 B nearest m/z: 110.0581 difference: 0.002299999999991087


In [61]:
matches = len(matched_intensities)

print("Matches:", matches)
print("Similarity:", matches / min(len(mzs_A), len(mzs_B)))

Matches: 9
Similarity: 0.05921052631578947


In [62]:
weighted_score = sum(matched_intensities) / np.sum(intensities_A)

print("Intensity-weighted similarity:", weighted_score)

Intensity-weighted similarity: 0.12768199936800761


In [63]:
def count_one_to_one_matches(mzs_A, mzs_B, tolerance=0.01):
    used_B = set()
    matches = 0

    for mz_A in mzs_A:
        distances = np.abs(mzs_B - mz_A)
        nearest = np.argmin(distances)

        if distances[nearest] <= tolerance and nearest not in used_B:
            matches += 1
            used_B.add(nearest)

    return matches

In [64]:
matches_1to1 = count_one_to_one_matches(mzs_A, mzs_B)

print("One-to-one matches:", matches_1to1)

One-to-one matches: 14


In [65]:
similarity_1to1 = matches_1to1 / min(len(mzs_A), len(mzs_B))

print("One-to-one similarity:", similarity_1to1)

One-to-one similarity: 0.09210526315789473


In [66]:
def weighted_one_to_one_similarity(mzs_A, intensities_A, mzs_B, tolerance=0.01):
    used_B = set()
    matched_intensity = 0.0

    for mz_A, intensity_A in zip(mzs_A, intensities_A):
        distances = np.abs(mzs_B - mz_A)
        nearest = np.argmin(distances)

        if distances[nearest] <= tolerance and nearest not in used_B:
            matched_intensity += intensity_A
            used_B.add(nearest)

    return matched_intensity / np.sum(intensities_A)

In [67]:
weighted_similarity_1to1 = weighted_one_to_one_similarity(
    mzs_A, intensities_A, mzs_B
)

print("Weighted one-to-one similarity:", weighted_similarity_1to1)

Weighted one-to-one similarity: 0.12768199936800761


In [72]:
rng = np.random.default_rng(42)

random_pairs = [
    tuple(rng.choice(len(train_sample), size=2, replace=False))
    for _ in range(100)
]

print(random_pairs[:5])

[(np.int64(11698), np.int64(101443)), (np.int64(57524), np.int64(56756)), (np.int64(91405), np.int64(11264)), (np.int64(12343), np.int64(69006)), (np.int64(96435), np.int64(99764))]


In [73]:
one_to_one_scores = []

for i, j in random_pairs:
    mz_A = train_sample.iloc[i]["ms2_mzs"]
    mz_B = train_sample.iloc[j]["ms2_mzs"]
    
    score = count_one_to_one_matches(mz_A, mz_B) / min(len(mz_A), len(mz_B))
    one_to_one_scores.append(score)

print("Mean one-to-one similarity:", np.mean(one_to_one_scores))

Mean one-to-one similarity: 0.079790632612349


In [74]:
weighted_scores = []

for i, j in random_pairs:
    mz_A = train_sample.iloc[i]["ms2_mzs"]
    int_A = train_sample.iloc[i]["ms2_normalized_intensities"]
    mz_B = train_sample.iloc[j]["ms2_mzs"]

    score = weighted_one_to_one_similarity(
        mz_A, int_A, mz_B
    )

    weighted_scores.append(score)

print("Mean weighted one-to-one similarity:", np.mean(weighted_scores))

Mean weighted one-to-one similarity: 0.06693824529335267


In [75]:
original_scores = []

for i, j in random_pairs:
    mz_A = train_sample.iloc[i]["ms2_mzs"]
    mz_B = train_sample.iloc[j]["ms2_mzs"]

    matches = count_peak_matches(mz_A, mz_B)
    score = matches / min(len(mz_A), len(mz_B))

    original_scores.append(score)

print("Mean original similarity:", np.mean(original_scores))

Mean original similarity: 0.10476068021488429


In [79]:
counts = train_sample["inchikey14"].value_counts()

molecule_id = counts[counts >= 2].index[0]

same_molecule = train_sample[
    train_sample["inchikey14"] == molecule_id
]

print("Molecule:", molecule_id)
print("Number of spectra:", len(same_molecule))

Molecule: AOYHVMRDIYZAQU
Number of spectra: 34


In [80]:
spectrum_A = same_molecule.iloc[0]
spectrum_B = same_molecule.iloc[1]

mzs_A = spectrum_A["ms2_mzs"]
intensities_A = spectrum_A["ms2_normalized_intensities"]

mzs_B = spectrum_B["ms2_mzs"]
intensities_B = spectrum_B["ms2_normalized_intensities"]

print("Spectrum A peaks:", len(mzs_A))
print("Spectrum B peaks:", len(mzs_B))

Spectrum A peaks: 142
Spectrum B peaks: 56


In [81]:
original = count_peak_matches(mzs_A, mzs_B) / min(len(mzs_A), len(mzs_B))

one_to_one = count_one_to_one_matches(mzs_A, mzs_B) / min(len(mzs_A), len(mzs_B))

weighted = weighted_one_to_one_similarity(
    mzs_A, intensities_A, mzs_B
)

print("Original similarity:", original)
print("One-to-one similarity:", one_to_one)
print("Weighted one-to-one similarity:", weighted)

Original similarity: 0.5178571428571429
One-to-one similarity: 0.4107142857142857
Weighted one-to-one similarity: 0.7874597565476691


In [82]:
different_molecule_id = counts[counts >= 2].index[1]

different_molecule = train_sample[
    train_sample["inchikey14"] == different_molecule_id
]

print("Different molecule:", different_molecule_id)
print("Number of spectra:", len(different_molecule))

Different molecule: AWRUAWMZJWGTJR
Number of spectra: 33


In [83]:
spectrum_C = different_molecule.iloc[0]

mzs_C = spectrum_C["ms2_mzs"]
intensities_C = spectrum_C["ms2_normalized_intensities"]

print("Spectrum C peaks:", len(mzs_C))

Spectrum C peaks: 37


In [84]:
same_molecule_score = weighted_one_to_one_similarity(
    mzs_A, intensities_A, mzs_B
)

different_molecule_score = weighted_one_to_one_similarity(
    mzs_A, intensities_A, mzs_C
)

print("Same molecule:", same_molecule_score)
print("Different molecule:", different_molecule_score)

Same molecule: 0.7874597565476691
Different molecule: 0.0


Weighted one-to-one peak matching produced substantially higher similarity for two spectra of the same molecule (0.7875) than for a spectrum from a different molecule (0.0000).

In [85]:
candidate_pool = train_sample.head(1000).copy()

print("Candidate spectra:", len(candidate_pool))

Candidate spectra: 1000


In [86]:
query = candidate_pool.iloc[0]

query_mzs = query["ms2_mzs"]
query_intensities = query["ms2_normalized_intensities"]

print("Query molecule:", query["inchikey14"])
print("Query peaks:", len(query_mzs))

Query molecule: AAALVYBICLMAMA
Query peaks: 52


In [87]:
all_mzs = np.concatenate(candidate_pool["ms2_mzs"].values)

bins = np.arange(
    np.floor(all_mzs.min()),
    np.ceil(all_mzs.max()) + 1,
    1.0
)

print("Number of bins:", len(bins) - 1)

Number of bins: 1727


In [89]:
query_vector = np.zeros(len(bins) - 1)

bin_indices = np.digitize(query_mzs, bins) - 1
bin_indices = np.clip(bin_indices, 0, len(query_vector) - 1)

for idx, intensity in zip(bin_indices, query_intensities):
    query_vector[idx] += intensity

print("Query vector length:", len(query_vector))
print("Non-zero bins:", np.count_nonzero(query_vector))

Query vector length: 1727
Non-zero bins: 48


In [90]:
from sklearn.metrics.pairwise import cosine_similarity

self_similarity = cosine_similarity(
    query_vector.reshape(1, -1),
    query_vector.reshape(1, -1)
)[0, 0]

print("Self cosine similarity:", self_similarity)

Self cosine similarity: 1.0


In [91]:
candidate = candidate_pool.iloc[1]

candidate_mzs = candidate["ms2_mzs"]
candidate_intensities = candidate["ms2_normalized_intensities"]

In [93]:
candidate_vector = np.zeros(len(bins) - 1)

bin_indices = np.digitize(candidate_mzs, bins) - 1
bin_indices = np.clip(bin_indices, 0, len(candidate_vector) - 1)


for idx, intensity in zip(bin_indices, candidate_intensities):
    candidate_vector[idx] += intensity

print("Candidate vector length:", len(candidate_vector))
print("Non-zero bins:", np.count_nonzero(candidate_vector))

Candidate vector length: 1727
Non-zero bins: 27


In [94]:
similarity = cosine_similarity(
    query_vector.reshape(1, -1),
    candidate_vector.reshape(1, -1)
)[0, 0]

print("Cosine similarity:", similarity)

Cosine similarity: 0.003704431663427373


In [95]:
similarities = []

for _, candidate in candidate_pool.iterrows():
    mzs = candidate["ms2_mzs"]
    intensities = candidate["ms2_normalized_intensities"]

    vector = np.zeros(len(bins) - 1)

    bin_indices = np.digitize(mzs, bins) - 1
    bin_indices = np.clip(bin_indices, 0, len(vector) - 1)

    for idx, intensity in zip(bin_indices, intensities):
        vector[idx] += intensity

    score = cosine_similarity(
        query_vector.reshape(1, -1),
        vector.reshape(1, -1)
    )[0, 0]

    similarities.append(score)

print("Number of similarity scores:", len(similarities))

Number of similarity scores: 1000


In [96]:
top_25_indices = np.argsort(similarities)[::-1][:25]

top_25_scores = np.array(similarities)[top_25_indices]

print(top_25_scores)

[1.         0.55221986 0.55171834 0.47389226 0.39324357 0.39298179
 0.3854962  0.36146191 0.34791264 0.34320579 0.33646032 0.32897411
 0.3243102  0.30940207 0.3061717  0.28392868 0.26019928 0.25320803
 0.25111818 0.23739043 0.23679238 0.22968085 0.17175468 0.16974419
 0.16084403]


In [97]:
top_25_results = candidate_pool.iloc[top_25_indices][
    ["inchikey14", "normalized_smiles"]
].copy()

top_25_results["similarity"] = top_25_scores

print(top_25_results.head(10))

         inchikey14                                  normalized_smiles  \
0    AAALVYBICLMAMA          O=C1NC(=O)c2cc(Nc3ccccc3)c(Nc3ccccc3)cc21   
849  IYRMWMYZSQPJKC            O=c1c(O)c(-c2ccc(O)cc2)oc2cc(O)cc(O)c12   
809  IQPNAANSBPBGFQ            O=c1cc(-c2ccc(O)c(O)c2)oc2cc(O)cc(O)c12   
352  DJOJDHGQRNZXQQ               COc1cc(O)c2c(c1)OC(c1ccc(O)cc1)CC2=O   
397  DSFGXPJYDCSWTA             CN(CCO)CC(O)Cn1cnc2c1c(=O)n(C)c(=O)n2C   
165  BMPDWHIDQYTSHX                     NC(=O)N1c2ccccc2CC(O)c2ccccc21   
461  FFGPTBGBLSHEPO                       NC(=O)N1c2ccccc2C=Cc2ccccc21   
691  HMJIYCCIJYRONP  COC(=O)C1=C(C)NC(C)=C(C(=O)OC(C)C)C1c1cccc2nonc12   
579  GNBHRKFJIUUOQI         O=C1OC2(c3ccc(O)cc3Oc3cc(O)ccc32)c2ccccc21   
690  HMJIYCCIJYRONP  COC(=O)C1=C(C)NC(C)=C(C(=O)OC(C)C)C1c1cccc2nonc12   

     similarity  
0      1.000000  
849    0.552220  
809    0.551718  
352    0.473892  
397    0.393244  
165    0.392982  
461    0.385496  
691    0.361462  
579    0.347913  
690  

In [98]:
retrieved = candidate_pool.iloc[top_25_indices].copy()

retrieved = retrieved[
    retrieved.index != candidate_pool.index[0]
]

print("Retrieved candidates:", len(retrieved))

Retrieved candidates: 24


In [99]:
unique_molecules = retrieved["inchikey14"].nunique()

print("Unique molecules:", unique_molecules)

Unique molecules: 23


In [100]:
retrieved_unique = retrieved.drop_duplicates(
    subset="inchikey14"
)

print("Unique candidate molecules:", len(retrieved_unique))

Unique candidate molecules: 23


In [102]:
retrieved = candidate_pool.iloc[top_25_indices].copy()
retrieved["similarity"] = top_25_scores

retrieved = retrieved[
    retrieved.index != candidate_pool.index[0]
]

In [103]:
print(retrieved.columns)

Index(['inchikey14', 'normalized_smiles', 'ms2_mzs',
       'ms2_normalized_intensities', 'similarity'],
      dtype='str')


In [104]:
retrieved_unique = (
    retrieved
    .sort_values("similarity", ascending=False)
    .drop_duplicates("inchikey14")
)

print("Unique molecules:", len(retrieved_unique))
print(retrieved_unique[["inchikey14", "similarity"]].head(10))

Unique molecules: 23
         inchikey14  similarity
849  IYRMWMYZSQPJKC    0.552220
809  IQPNAANSBPBGFQ    0.551718
352  DJOJDHGQRNZXQQ    0.473892
397  DSFGXPJYDCSWTA    0.393244
165  BMPDWHIDQYTSHX    0.392982
461  FFGPTBGBLSHEPO    0.385496
691  HMJIYCCIJYRONP    0.361462
579  GNBHRKFJIUUOQI    0.347913
714  HSAOETBFVAWNRP    0.336460
829  IVTVGDXNLFLDRM    0.328974


In [106]:
true_molecule = query["inchikey14"]

print("True molecule:", true_molecule)
print(
    "Present in retrieved candidates:",
    true_molecule in retrieved_unique["inchikey14"].values
)

True molecule: AAALVYBICLMAMA
Present in retrieved candidates: False


In [107]:
queries = candidate_pool.head(10).copy()

print("Number of queries:", len(queries))

Number of queries: 10


In [108]:
def retrieve_top25(query, candidate_pool, bins):
    query_mzs = query["ms2_mzs"]
    query_intensities = query["ms2_normalized_intensities"]

    # Create query vector
    query_vector = np.zeros(len(bins) - 1)

    indices = np.digitize(query_mzs, bins) - 1
    indices = np.clip(indices, 0, len(query_vector) - 1)

    for idx, intensity in zip(indices, query_intensities):
        query_vector[idx] += intensity

    similarities = []

    # Compare against candidates
    for _, candidate in candidate_pool.iterrows():
        vector = np.zeros(len(bins) - 1)

        indices = np.digitize(candidate["ms2_mzs"], bins) - 1
        indices = np.clip(indices, 0, len(vector) - 1)

        for idx, intensity in zip(
            indices,
            candidate["ms2_normalized_intensities"]
        ):
            vector[idx] += intensity

        score = cosine_similarity(
            query_vector.reshape(1, -1),
            vector.reshape(1, -1)
        )[0, 0]

        similarities.append(score)

    # Top 25 spectra
    top_indices = np.argsort(similarities)[::-1][:25]

    retrieved = candidate_pool.iloc[top_indices].copy()
    retrieved["similarity"] = np.array(similarities)[top_indices]

    # Remove query itself
    retrieved = retrieved[
        retrieved.index != query.name
    ]

    # Molecule-level deduplication
    retrieved = (
        retrieved
        .sort_values("similarity", ascending=False)
        .drop_duplicates("inchikey14")
    )

    return retrieved.head(25)

In [109]:
result = retrieve_top25(
    queries.iloc[0],
    candidate_pool,
    bins
)

print(result[["inchikey14", "similarity"]].head())
print("Retrieved molecules:", len(result))

         inchikey14  similarity
849  IYRMWMYZSQPJKC    0.552220
809  IQPNAANSBPBGFQ    0.551718
352  DJOJDHGQRNZXQQ    0.473892
397  DSFGXPJYDCSWTA    0.393244
165  BMPDWHIDQYTSHX    0.392982
Retrieved molecules: 23


In [110]:
true_molecule = queries.iloc[0]["inchikey14"]

print("True molecule:", true_molecule)
print(
    "Retrieved:",
    true_molecule in result["inchikey14"].values
)

True molecule: AAALVYBICLMAMA
Retrieved: False


In [111]:
all_predictions = []
true_molecules = []

for _, query in queries.iterrows():

    result = retrieve_top25(
        query,
        candidate_pool,
        bins
    )

    predictions = result["inchikey14"].tolist()

    all_predictions.append(predictions)
    true_molecules.append(query["inchikey14"])

print("Queries evaluated:", len(all_predictions))
print("Predictions for first query:", len(all_predictions[0]))

Queries evaluated: 10
Predictions for first query: 23


In [112]:
score = mrr_at_25(
    all_predictions,
    true_molecules
)

print("MRR@25:", score)

MRR@25: 0.0


In [113]:
for molecule in true_molecules:
    print(
        molecule,
        molecule in candidate_pool["inchikey14"].values
    )

AAALVYBICLMAMA True
AAKJLRGGTJKAMG True
AANFHDFOMFRLLR True
AAOVKJBEBIDNHE True
ACFIXJIJDZMPPO True
ACGDKVXYNVEAGU True
ACNHBCIZLNNLRS True
ACPOUJIDANTYHO True
ACRHBAYQBXXRTO True
ACTIUHUUMQJHFO True


In [114]:
query = queries.iloc[0]

scores = []

for _, candidate in candidate_pool.iterrows():
    # same vectorization code
    vector = np.zeros(len(bins) - 1)

    indices = np.digitize(
        candidate["ms2_mzs"], bins
    ) - 1

    indices = np.clip(indices, 0, len(vector) - 1)

    for idx, intensity in zip(
        indices,
        candidate["ms2_normalized_intensities"]
    ):
        vector[idx] += intensity

    score = cosine_similarity(
        query_vector.reshape(1, -1),
        vector.reshape(1, -1)
    )[0, 0]

    scores.append(score)

ranked_indices = np.argsort(scores)[::-1]

true_rows = candidate_pool[
    candidate_pool["inchikey14"] == query["inchikey14"]
]

true_ranks = [
    np.where(ranked_indices == idx)[0][0] + 1
    for idx in true_rows.index
]

print("True molecule spectrum ranks:", true_ranks)

True molecule spectrum ranks: [np.int64(1)]


In [115]:
molecule_counts = candidate_pool["inchikey14"].value_counts()

multi_spectrum_molecules = molecule_counts[
    molecule_counts >= 2
]

print("Molecules with 2+ spectra:", len(multi_spectrum_molecules))
print(multi_spectrum_molecules.head())

Molecules with 2+ spectra: 2
inchikey14
HMJIYCCIJYRONP    2
HYIMSNHJOBLJNT    2
Name: count, dtype: int64


In [116]:
candidate_pool = train_sample.head(10000).copy()

print("Candidate spectra:", len(candidate_pool))

Candidate spectra: 10000


In [117]:
molecule_counts = candidate_pool["inchikey14"].value_counts()

multi_spectrum_molecules = molecule_counts[
    molecule_counts >= 2
]

print("Molecules with 2+ spectra:", len(multi_spectrum_molecules))
print(multi_spectrum_molecules.head(10))

Molecules with 2+ spectra: 1158
inchikey14
ACAIOSQAEAUJLZ    22
ABAWQKPDKUSIBA    21
ABYCMWGQUVUISN    20
ABIIVYZPHUTRHZ    18
ACJDNOONJXDRDT    18
ACLOLZVKLOFXDX    18
ABXFAELWDNMNBN    17
ACYCONYTQZHVPT    17
ADSCTDGNVOOLMR    17
AAOVCEIQPDDKAC    16
Name: count, dtype: int64


In [118]:
valid_queries = (
    candidate_pool[
        candidate_pool["inchikey14"].isin(multi_spectrum_molecules.index)
    ]
    .groupby("inchikey14", group_keys=False)
    .head(1)
    .head(10)
)

print("Valid queries:", len(valid_queries))
print(valid_queries["inchikey14"].tolist())

Valid queries: 10
['HMJIYCCIJYRONP', 'HYIMSNHJOBLJNT', 'RZTAMFZIAATZDJ', 'UIAGMCDKSXEBJQ', 'WDZVGELJXXEGPV', 'ZBBHBTPTTSWHBA', 'AAAAMGLVZZXVDT', 'AAAILXDLOQPKFL', 'AAAJHMPMZMRLLC', 'AAAUDFYBEREPID']


In [119]:
all_predictions = []
true_molecules = []

for _, query in valid_queries.iterrows():
    result = retrieve_top25(
        query,
        candidate_pool,
        bins
    )

    predictions = result["inchikey14"].tolist()

    all_predictions.append(predictions)
    true_molecules.append(query["inchikey14"])

print("Queries evaluated:", len(all_predictions))
print("Predictions for first query:", len(all_predictions[0]))

Queries evaluated: 10
Predictions for first query: 14


In [120]:
score = mrr_at_25(
    all_predictions,
    true_molecules
)

print("MRR@25:", score)

MRR@25: 0.31


In [121]:
for i, (preds, true_molecule) in enumerate(
    zip(all_predictions, true_molecules)
):
    if true_molecule in preds:
        rank = preds.index(true_molecule) + 1
        print(i, true_molecule, "→ Rank:", rank)
    else:
        print(i, true_molecule, "→ Not in top 25")

0 HMJIYCCIJYRONP → Rank: 1
1 HYIMSNHJOBLJNT → Not in top 25
2 RZTAMFZIAATZDJ → Not in top 25
3 UIAGMCDKSXEBJQ → Not in top 25
4 WDZVGELJXXEGPV → Not in top 25
5 ZBBHBTPTTSWHBA → Not in top 25
6 AAAAMGLVZZXVDT → Not in top 25
7 AAAILXDLOQPKFL → Rank: 1
8 AAAJHMPMZMRLLC → Rank: 1
9 AAAUDFYBEREPID → Rank: 10
